packages/react-native-gesture-handler/shared/RNGHRuntimeDecorator.cpp

```diff
@@ -0,0 +1,124 @@
+#ifndef ANDROID
+#include <react/renderer/components/text/ParagraphShadowNode.h>
+#include <react/renderer/components/text/TextShadowNode.h>
+#endif
+
+#include <react/renderer/uimanager/primitives.h>
+
+#include "RNGHRuntimeDecorator.h"
+
+namespace gesturehandler {
+
+using namespace facebook;
+using namespace facebook::react;
+
+void RNGHRuntimeDecorator::installRNRuntimeBindings(
+    jsi::Runtime &rnRuntime,
+    std::function<void(int, int)> &&setGestureState) {
+  const auto isViewFlatteningDisabled = jsi::Function::createFromHostFunction(
+      rnRuntime,
+      jsi::PropNameID::forAscii(rnRuntime, "_isViewFlatteningDisabled"),
+      1,
+      [](jsi::Runtime &runtime,
+         const jsi::Value &,
+         const jsi::Value *args,
+         size_t argumentCount) -> jsi::Value {
+        if (!args[0].isObject()) {
+          return jsi::Value::null();
+        }
+
+        const auto shadowNode = shadowNodeFromValue(runtime, args[0]);
+
+#ifndef ANDROID
+        if (dynamic_pointer_cast<const ParagraphShadowNode>(shadowNode)) {
+          return jsi::Value(true);
+        }
+
+        if (dynamic_pointer_cast<const TextShadowNode>(shadowNode)) {
+          return jsi::Value(true);
+        }
+#endif
+
+        const auto isFormsStackingContext = shadowNode->getTraits().check(
+            ShadowNodeTraits::FormsStackingContext);
+
+        // This is done using component names instead of type checking because
+        // of duplicate symbols for RN types, which prevent RTTI from working.
+        const auto &componentName = shadowNode->getComponentName();
+        const auto isTextOrParagraphComponent =
+            strcmp(componentName, "Paragraph") == 0 ||
+            strcmp(componentName, "Text") == 0;
+
+        return jsi::Value(isFormsStackingContext || isTextOrParagraphComponent);
+      });
+
+  rnRuntime.global().setProperty(
+      rnRuntime,
+      "_isViewFlatteningDisabled",
+      std::move(isViewFlatteningDisabled));
+
+  auto setGestureStateAsync = jsi::Function::createFromHostFunction(
+      rnRuntime,
+      jsi::PropNameID::forAscii(rnRuntime, "_setGestureStateAsync"),
+      2,
+      [setGestureState](
+          jsi::Runtime &rt,
+          const jsi::Value &,
+          const jsi::Value *args,
+          size_t argumentCount) -> jsi::Value {
+        if (argumentCount == 2) {
+          const auto handlerTag = static_cast<int>(args[0].asNumber());
+          const auto state = static_cast<int>(args[1].asNumber());
+
+          setGestureState(handlerTag, state);
+        }
+        return jsi::Value::undefined();
+      });
+
+  rnRuntime.global().setProperty(
+      rnRuntime, "_setGestureStateAsync", std::move(setGestureStateAsync));
+}
+
+bool RNGHRuntimeDecorator::installUIRuntimeBindings(
+    jsi::Runtime &rnRuntime,
+    std::function<void(int, int)> &&setGestureState) {
+  const auto runtimeHolder =
+      rnRuntime.global().getProperty(rnRuntime, "_WORKLET_RUNTIME");
```

Comment:
Do I get it correctly that this is added to rnRuntime by Reanimated? From which version of lib is it there? Also, can't we have some race conditions that it is not yet there when you call this method or something like that? Btw, is it safe to have some kind of pointer from one runtime to another?

Reply:
Just for the context, `_WORKLET_RUNTIME` was added a few years ago to support integration of Reanimated with Expo GL.